### PR TITLE
Add option publish_hz to publish current rate.

### PR DESCRIPTION
**Pull request by dornhege**
This makes it easy to use this in other tools such as rqt_plot.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
Wouldn't it be better to use the Python API to get the current publish rate rather then publishing that information on another topic?

**Comment by dornhege** [comment]
This would require someone to write code to use this. In this way, one can just start rostopic hz + rqt_plot and plot the current rate.

File: tools/rostopic/src/rostopic/__init__.py [hunk 1]
```diff
@@ -62,6 +62,7 @@
 import rosgraph
 #TODO: lazy-import rospy or move rospy-dependent routines to separate location
 import rospy
+import std_msgs
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py:65]: This new dependency needs to be listed in the package manifest.

File: tools/rostopic/src/rostopic/__init__.py [hunk 2]
```diff
@@ -106,12 +107,12 @@ def __init__(self, window_size, filter_expr=None, use_wtime=False):
         self.times =[]
         self.filter_expr = filter_expr
         self.use_wtime = use_wtime
-        
+
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py]: Please avoid any unrelated whitespace changes.

Several below.

File: tools/rostopic/src/rostopic/__init__.py [hunk 3]
```diff
@@ -1297,10 +1307,13 @@ def _rostopic_cmd_hz(argv):
     parser.add_option("--wall-time",
                       dest="use_wtime", default=False, action="store_true",
                       help="calculates rate using wall time which can be helpful when clock isnt published during simulation")
+    parser.add_option("--publish-hz",
+                    dest="publish_hz", default=False, action="store_true",
+                    help="publishes the rate on ~hz to be used by other program (e.g. for plotting)")
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py]: Would it make sense to pass a topic name here? Otherwise if the user runs multiple of these commands it becomes difficult to choose the "right" one.

File: tools/rostopic/src/rostopic/__init__.py [hunk 4]
```diff
@@ -210,7 +218,9 @@ def _rostopic_hz(topic, window_size=-1, filter_expr=None, use_wtime=False):
     while not rospy.is_shutdown():
         _sleep(1.0)
         rt.print_hz()
-
+        if pub:
+            rate, min_delta, max_delta, std_dev, n = rt.compute_hz()
+            pub.publish(rate)
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py]: Would it make sense to publish more than just the rate? Maybe the user wants to plot the other values too.

File: tools/rostopic/src/rostopic/__init__.py [hunk 5]
```diff
@@ -145,25 +146,17 @@ def callback_hz(self, m):
...
-    def print_hz(self):
-        """
-        print the average publishing rate to screen
-        """
-        if not self.times:
-            return
-        elif self.msg_tn == self.last_printed_tn:
-            print("no new messages")
-            return
+    def compute_hz(self):
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py]: Please add a docblock for this function.

**Review by dirk-thomas — COMMENTED** [review]
Now I understand the use case. That would be a good feature. I added a few comments inline.

Please also change the targeted branch of the PR to be `lunar-devel` (you can do that by clicking the edit button beside the title but might need to ensure that your changes apply or are based ontop of that branch).

**Comment by dornhege** [comment]
@dirk-thomas I addressed the minor issues. Not sure, what you meant with "the other values". Min/max, etc. could also be published. For now, I added the same kind of option for band width. A merge into lunar will require additional changes as the implementation has changed.

**Comment by dirk-thomas** [comment]
I was more thinking about publishing all printed values in a single message. E.g. `rostopic hz --publish /foo` would publish a `Hz` message contains all fields usually printed: `rate`, `min_delta`, `max_delta`, `std_dev`, and (maybe) `n`. That way the user can plot any of the values and is not limited to `rate`.

Please update the PR to target `lunar-devel`. Patches are by default only accepted into the latest branch (except if they fix a bug only present in older distros).

**Comment by dornhege** [comment]
Closing this in favor of #1072 for lunar.
